src/libcore/fmt/num.rs

```diff
@@ -256,6 +257,161 @@ macro_rules! impl_Display {
     };
 }
 
+macro_rules! impl_Exp {
+    ($($t:ident),* as $u:ident via $conv_fn:ident named $name:ident) => {
+        fn $name(
+            mut n: $u,
+            is_nonnegative: bool,
+            upper: bool,
+            f: &mut fmt::Formatter<'_>
+        ) -> fmt::Result {
+            let mut exponent = 0;
+            // count and remove trailing decimal zeroes
+            while n % 10 == 0 && n >= 10 {
+                n /= 10;
+                exponent += 1;
+            }
+            let trailing_zeros = exponent;
+
+            let (added_precision, subtracted_precision) = match f.precision() {
+                Some(fmt_prec) => {
+                    // number of decimal digits minus 1
+                    let prec = {
+                        let mut tmp = n;
+                        let mut count = 0;
+                        while tmp >= 10 {
+                            tmp /= 10;
+                            count += 1;
+                        }
+                        count
+                    };
+                    (fmt_prec.saturating_sub(prec), prec.saturating_sub(fmt_prec))
+                }
+                None => (0,0)
+            };
+            for _ in 1..subtracted_precision {
+                n/=10;
+                exponent += 1;
+            }
+            if subtracted_precision != 0 {
+                let rem = n % 10;
+                n /= 10;
+                exponent += 1;
+                // round up last digit
+                if rem >= 5 {
+                    n += 1;
+                }
+            }
+
+            // 39 digits (worst case u128) + . = 40
+            let mut buf = [MaybeUninit::<u8>::uninit(); 40];
+            let mut curr = buf.len() as isize; //index for buf
+            let buf_ptr = MaybeUninit::first_ptr_mut(&mut buf);
+            let lut_ptr = DEC_DIGITS_LUT.as_ptr();
+
+            // decode 2 chars at a time
+            while n >= 100 {
+                let d1 = ((n % 100) as isize) << 1;
+                n /= 100;
```

Comment:
Let's order the update to `n` and `exponent` and the writing to the buffer and updating `curr` the same way in each of these blocks.